developer_manual/client_apis/OCS/ocs-openapi.rst

```diff
@@ -0,0 +1,329 @@
+===================
+OCS OpenAPI support
+===================
+
+This page explains you how you can add OpenAPI support to your app such that you can automatically generated an OpenAPI specification for your server code.
+
+Please read the whole tutorial before starting to adjust your app.
+
+Don't be afraid that you don't know everything from the start.
+The openapi-extractor tool gives you many warnings and fails if there is something utterly broken that would not work.
+Just let it run, it will tell you if there is something wrong.
+Psalm will also help you validate your changes.
+
+Requirements and prerequisites
+------------------------------
+
+- Psalm has been setup for the app
+- The app supports at least Nextcloud 28
+- Your APIs are exposed via OCS (see :ref:`OCS <ocscontroller>`). It is possible to also expose APIs in different ways, but those will not be explained here because they are considered legacy.
+
+Best practices
+--------------
+
+- Be as explicit and concrete as possible about types. The closer you narrow down a type without violating any constraints the better the resulting specification will be.
+- For empty data you should always use `null` if possible. If your API is well designed it should always work. If for some reason you can not do that, you should use `new \stdClass()` and not `[]`. The problem with `[]` in PHP is that will be converted into `[]` in JSON, but you want to have `{}` in JSON if the data you return is empty. For empty arrays that should be arrays it is not a problem.
```

Comment:
I can see people not knowing the issue getting confused here.

The second part starting with `The problem with ...` has a lot of inline code where words might be better.
It's not the best but maybe something like:

> Keep in mind that JSON distinguishes between empty lists and empty objects so a PHP `[]` should only be used to indicate an array with no items.

Reply:
Is the new version better?